File: VideoReadWriteDisplay/videoRead.cpp
#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <cstdint>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>

int main(int argc, char** argv) {
    const std::string keys =
        "{help h usage ?||Show this help message}"
        "{input|chaplin.mp4|Input video path}"
        "{display||Show frames in a window}"
        "{max-frames|-1|Maximum frames; -1 reads all}";
    cv::CommandLineParser parser(argc, argv, keys);
    if (parser.has("help")) {
        parser.printMessage();
        return 0;
    }
    if (!parser.check()) {
        parser.printErrors();
        return 2;
    }

    try {
        const std::string inputPath = parser.get<std::string>("input");
        const bool display = parser.has("display");
        const int maxFrames = parser.get<int>("max-frames");
        if (maxFrames == 0 || maxFrames < -1) {
            throw std::invalid_argument("max-frames must be positive or -1.");
        }

        cv::VideoCapture capture(inputPath);
        if (!capture.isOpened()) {
            throw std::runtime_error("Could not open input video: " + inputPath);
        }

        cv::Mat frame;
        int framesRead = 0;
        std::uint64_t firstFrameChecksum = 0;
        while ((maxFrames < 0 || framesRead < maxFrames) &&
               capture.read(frame) && !frame.empty()) {
            ++framesRead;
            if (framesRead == 1) {
                firstFrameChecksum =
                    static_cast<std::uint64_t>(cv::sum(frame)[0]) +
                    static_cast<std::uint64_t>(cv::sum(frame)[1]) +
                    static_cast<std::uint64_t>(cv::sum(frame)[2]);
            }
            if (display) {
                cv::imshow("Video", frame);
                const int key = cv::waitKey(25) & 0xff;
                if (key == 27 || key == 'q') {
                    break;
                }
            }
        }
        if (framesRead == 0) {
            throw std::runtime_error("No decodable frames found.");
        }

        std::cout << std::fixed << std::setprecision(4)
                  << "{\"frames_read\":" << framesRead
                  << ",\"frame_size\":["
                  << static_cast<int>(capture.get(cv::CAP_PROP_FRAME_WIDTH))
                  << ','
                  << static_cast<int>(capture.get(cv::CAP_PROP_FRAME_HEIGHT))
                  << "],\"fps\":" << capture.get(cv::CAP_PROP_FPS)
                  << ",\"reported_frame_count\":"
                  << static_cast<int>(capture.get(cv::CAP_PROP_FRAME_COUNT))
                  << ",\"first_frame_checksum\":" << firstFrameChecksum
                  << "}\n";
        return 0;
    } catch (const cv::Exception& error) {
        std::cerr << "OpenCV error: " << error.what() << '\n';
    } catch (const std::exception& error) {
        std::cerr << "error: " << error.what() << '\n';
    }
    return 2;
}

File: VideoReadWriteDisplay/videoRead.py
#!/usr/bin/env python3
"""Read a video safely, optionally display it, and report measured metadata."""

from __future__ import annotations

import argparse
import json
from pathlib import Path
from typing import Any

import cv2


def read_video(
    input_path: Path,
    *,
    display: bool = False,
    max_frames: int | None = None,
) -> dict[str, Any]:
    if max_frames is not None and max_frames <= 0:
        raise ValueError("max_frames must be positive")

    capture = cv2.VideoCapture(str(input_path))
    if not capture.isOpened():
        raise FileNotFoundError(f"Could not open input video: {input_path}")

    frames_read = 0
    first_frame_checksum = None
    try:
        while max_frames is None or frames_read < max_frames:
            ok, frame = capture.read()
            if not ok or frame is None:
                break
            frames_read += 1
            if first_frame_checksum is None:
                first_frame_checksum = int(frame.astype("uint64").sum())
            if display:
                cv2.imshow("Video", frame)
                if cv2.waitKey(25) & 0xFF in (27, ord("q")):
                    break

        if frames_read == 0:
            raise RuntimeError(f"No decodable frames found in: {input_path}")
        return {
            "frames_read": frames_read,
            "frame_size": [
                int(capture.get(cv2.CAP_PROP_FRAME_WIDTH)),
                int(capture.get(cv2.CAP_PROP_FRAME_HEIGHT)),
            ],
            "fps": capture.get(cv2.CAP_PROP_FPS),
            "reported_frame_count": int(
                capture.get(cv2.CAP_PROP_FRAME_COUNT)
            ),
            "first_frame_checksum": first_frame_checksum,
        }
    finally:
        capture.release()
        if display:
            cv2.destroyAllWindows()


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--input",
        type=Path,
        default=Path(__file__).parent / "chaplin.mp4",
    )
    parser.add_argument("--display", action="store_true")
    parser.add_argument("--max-frames", type=int)
    return parser


def main() -> int:
    args = build_parser().parse_args()
    try:
        summary = read_video(
            args.input,
            display=args.display,
            max_frames=args.max_frames,
        )
    except (FileNotFoundError, RuntimeError, ValueError) as exc:
        print(f"error: {exc}")
        return 2
    print(json.dumps(summary, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
